### PR TITLE
spack install : added --log-format option (incorporates test-install command)

### LINKED ISSUE #1907: Merge commands with overlapping responsibilities
I was wondering if anybody else was in favor of merging :

- [x] `spack package-list` in `spack list`
- [x] `spack test-install` in `spack install`

and demote them from full commands to flags in another command. If there's agreement I'll start working on it.

### PR DESCRIPTION
**Pull request by alalazo**
##### Modifications:
- [x] fixes #1907
- [x] adds two options to install : `--log-format` and `--log-file`
- [x] default log file is written in `join_path(spack.var_path, 'junit-report')` instead of `join_path(os.getcwd(), 'junit-report')`

### CODE REVIEW

File: lib/spack/spack/cmd/install.py [hunk 1]
```diff
@@ -71,7 +80,207 @@ def setup_parser(subparser):
...
+                # If already installed set the spec as skipped
+                start_time = time.time()
+                # PackageBase.do_install
+                func(self, *args, **kwargs)
+                duration = time.time() - start_time
+                test_case.set_duration(duration)
+                test_case.set_result(TestResult.PASSED)
+            except InstallError:
+                # Check if the package relies on dependencies that
+                #  did not install
```
Comment (davydden) [inline lib/spack/spack/cmd/install.py]: extra whitespace

**Comment by alalazo** [comment]
@scheibelp  I was wondering if you could review the changes here and see if anything looks wrong.

@adamjstewart @tgamblin @davydden @citibeth Is there any other information that you would like to see in the JunitXML report and is not there yet ?

**Comment by alalazo** [comment]
Ok, giving the command: 
```
spack install --log-format=junit  netcdf
```
with only the root that needs to be built produces the following report in `var/spack/junit-report`:
```xml
<?xml version="1.0" ?>
<testsuite errors="0" failures="0" tests="11">
        <testcase classname="zlib" name="zlib@1.2.8%gcc@6.2.0 arch=linux-Ubuntu14-x86_64-asydrba" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="openssl" name="openssl@1.0.2j%gcc@6.2.0 arch=linux-Ubuntu14-x86_64-hibnfda" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="curl" name="curl@7.50.3%gcc@6.2.0 arch=linux-Ubuntu14-x86_64-2ffacqm" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="libsigsegv" name="libsigsegv@2.10%gcc@6.2.0 arch=linux-Ubuntu14-x86_64-g3qpmbi" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="m4" name="m4@1.4.17%gcc@6.2.0+sigsegv arch=linux-Ubuntu14-x86_64-lhgqa6s" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="pkg-config" name="pkg-config@0.29.1%gcc@6.2.0 arch=linux-Ubuntu14-x86_64-n6bkle3" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="libpciaccess" name="libpciaccess@0.13.4%gcc@6.2.0 arch=linux-Ubuntu14-x86_64-eo2siet" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="hwloc" name="hwloc@1.11.4%gcc@6.2.0 arch=linux-Ubuntu14-x86_64-3ostwel" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="openmpi" name="openmpi@2.0.1%gcc@6.2.0~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm~verbs+vt arch=linux-Ubuntu14-x86_64-s3qbtby" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="hdf5" name="hdf5@1.10.0-patch1%gcc@6.2.0+cxx~debug+fortran+mpi+shared~szip~threadsafe arch=linux-Ubuntu14-x86_64-3a22tnq" time="0.0">
                <skipped message="Skipped [already installed]" type="already_installed"/>
        </testcase>
        <testcase classname="netcdf" name="netcdf@4.4.1%gcc@6.2.0~hdf4 maxdims=1024 maxvars=8192+mpi arch=linux-Ubuntu14-x86_64-x7rkaqv" time="36.8761630058"/>
</testsuite>
```

**Comment by davydden** [comment]
@alalazo thanks, looks good. I don't have in mind any further information we would need.

Supposedly, there could be a script somewhere which can parse numerous XML (from different users who contribute test results) into a single output attached to a release? Maybe a document or something that could be easy to navigate for a user to see which packages are tested on which architectures and which compilers for that particular release.

**Comment by davydden** [comment]
p.s. It do not suggest to do this within the current PR, just throwing ideas on the table.

**Comment by alalazo** [comment]
@davydden To be fair @scheibelp implemented JUnitXML in the first place. I am just refactoring the stuff to reduce the number of similar commands in Spack without loosing the functionality. That said I think it will be an essential building block for a release installation test. 

What we are currently missing for that (and won't likely happen before SC) is imo:
- a file format to describe a matrix of things to be try-installed for the release
- the logic to trigger **al**l those tests and create a report
- a dashboard online that can display JUnitXML submitted from the various groups

I know there are some discussions that touched these topics in latest issues, but couldn't find them right now...

**Comment by davydden** [comment]
> To be fair @scheibelp implemented JUnitXML in the first place. I am just refactoring the stuff to reduce the number of similar commands in Spack without loosing the functionality

I understand that, i was trying to see how would one later use this functionality for testing.

> a dashboard online that can display JUnitXML submitted from the various groups

essentially i was speaking about this. Something like CDash would be good.

> I know there are some discussions that touched these topics in latest issues, but couldn't find them right now...

One of them is https://github.com/LLNL/spack/issues/2014 , i suppose...

**Comment by davydden** [comment]
@alalazo : I think there is one more thing we have to log -- hash of the Spack and maybe the most recent tag. Maybe within `testsuite errors="0" failures="0" tests="11" hash="xyz" tag="abc"`.

**Comment by tgamblin** [comment]
So, I could set up a CDash instance in EC2 so we could at least start reporting test results.  It apparently [has some support NUnit format](http://www.vtk.org/Wiki/CDash:XML#CDash_with_JUnit) so we could start by adding a few fields to the current test format. What do people think of that?

I am worried that the matrix display stuff is going to require more infrastructure than we have.  I think Spack ultimately needs its own dashboard of some sort, where we could do `spack find`-ish queries on test results and query a server-side database of `spec.yaml`s.  But that's not happening before SC.

**Comment by citibeth** [comment]
What do I do with this xml file?  It looks interesting but is not really
readable without further processing.

**Comment by tgamblin** [comment]
@citibeth: you submit it to Bamboo, Jenkins, or CDash.

**Comment by alalazo** [comment]
@citibeth All major CI framework supports that to display results in various way.

**Comment by davydden** [comment]
@alalazo :  you probably missed my comment above, but do we need to encode Spack's hash used to create tests? I would say yes.

**Comment by alalazo** [comment]
@davydden I didn't miss the message. I just need to have a look at the NUnit format to see how they support writing metadata (if they do at all).

**Comment by davydden** [comment]
i see, no worries.

**Comment by citibeth** [comment]
This will be pretty awesome.

**Comment by tgamblin** [comment]
@alalazo: do you want this merged now to update NUnit later, or do you want to keep it open?

**Comment by alalazo** [comment]
@tgamblin Fine with me to merge this and update later

**Comment by hegner** [comment]
With #2433 the minimal requirements for uploading to CDash are now there. See https://cdash.cern.ch/index.php?project=Spack

For now, that dashboard is free for submission via e.g.
`curl -k --upload-file var/spack/junit-report/test-zlib-1.2.8-pfy5ppjohs2f7cmt3br4bozg7rd4zack.xml https://cdash.cern.ch/submit.php?project=Spack`

**Pull request merged**